### PR TITLE
Added proper focus handling when pushing and popping routes

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

The proposed change will change focus handling when pushing and popping routes so that the `FocusScopeNode` for the route receives focus when pushed, and that the `FocusScopeNode` in the navigator receives focus when the route is popped.

This means that the last `setFirstFocus` call on the scope is used to determine which control actually receives focus. When the focus scope receives focus, it traverses its children, trying to find a non-scope node that is the "first focus" of itself or a child node.

This is a breaking change, because the focus behavior has changed. If you push a route after this change, and had a 'first focus' set on a widget via `FocusScopeNode.setFirstFocus`, it won't currently receive focus immediately, but after this change it will. Similarly, if you pop a route after this change, the focus will go back to where it was before the route was pushed, which is correct, but different from what happens now.

## Related Issues

Addresses #13264

## Tests

- Added tests for route pushing/popping.
- Adjusted other tests to account for new behavior.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.
- [X] I checked all the boxes!

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [X] Yes, this is a breaking change. ([Breaking change announcement](https://groups.google.com/forum/#!topic/flutter-announce/BjPKnOw5jzU))

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/routes.dart [hunk 1]
```diff
@@ -629,11 +630,16 @@ class _ModalScopeState<T> extends State<_ModalScope<T>> {
...
+    if (widget.route.isCurrent) {
+      widget.route.navigator.focusScopeNode.setFirstFocus(focusScopeNode);
+    }
     setState(fn);
   }
 
   @override
   Widget build(BuildContext context) {
+    final bool isReversing = widget.route.animation?.status == AnimationStatus.reverse;
+    focusScopeNode.canRequestFocus = !isReversing;
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/routes.dart]: What if we're in the middle of an iOS-back-swipe-route-dismiss gesture, i.e. what if route.navigator.userGestureInProgress is true?  Seems like we wouldn't want to take the focus then either?

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 2]
```diff
@@ -625,18 +632,28 @@ class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
...
-      final FocusScopeNode scope = enclosingScope;
-      assert(scope != null, 'Node has primary focus, but no enclosingScope.');
-      scope._focusedChildren.remove(this);
-      _manager?._willUnfocusNode(this);
+  ///
+  /// if [focusPrevious] is true, then rather than losing all focus, the focus
+  /// will be moved to the node that the [enclosingScope] thinks should have it,
+  /// based on its history of nodes that were set as first focus on it using
+  /// [FocusScopeNode.setFirstFocus].
+  void unfocus({bool focusPrevious = false}) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/focus_manager.dart]: ({ bool focusPrevious })

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 3]
```diff
@@ -133,7 +133,12 @@ abstract class Route<T> {
   /// The [didChangeNext] and [didChangePrevious] methods are typically called
   /// immediately after this method is called.
   @protected
-  TickerFuture didPush() => TickerFuture.complete();
+  @mustCallSuper
+  TickerFuture didPush() {
+    return TickerFuture.complete()..then<void>((void _) {
+      navigator.focusScopeNode.requestFocus();
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/navigator.dart]: This is safe even if more than one route is pushed before the futures complete?

**Review by HansMuller — APPROVED** [review]
I can't say that I understand all of the implications of this (sensible) change. LGTM though.

**Comment by ThinkDigitalSoftware** [comment]
@gspencergoog This causes an issue when the route that's pushed is not a modal route, like a dropdown button or a popupmenu. It goes against the standard android behavior.

**Comment by gspencergoog** [comment]
@ThinkDigitalSoftware  Can you elaborate?  This change should only change focus when pushing a route if `setFirstFocus` has been called on the `FocusScope` (or `autofocus` has been set on a `Focus` widget).  It doesn't focus the focus nodes directly, only their scopes. It's necessary to be able to do this in order for keyboard focus traversal to have a starting scope for focus.

**Comment by ThinkDigitalSoftware** [comment]
Well in my case, I haven't set any of those properties. I have a barebones example where when you tap a textfield and then a Dropdown, once the Dropdown is dismissed, the textfield is activated again and the keyboard always comes up, whether it was up previously or not. It makes sense for modal routes, because you're returning to where you are. But if you're filling out a form, this causes the screen to scroll back to the textfield and activate it, regardless of how far it is from the Dropdown /popup

**Comment by ThinkDigitalSoftware** [comment]
No

**Comment by ThinkDigitalSoftware** [comment]
This pr was meant to address some of the issues, but not all https://github.com/flutter/flutter/pull/41574

**Comment by ThinkDigitalSoftware** [comment]
ezgif-4-408d89423126

**Comment by gspencergoog** [comment]
@ThinkDigitalSoftware, I see the issue.  I've also seen something else that is related just now, so I'm filing a bug and will look at it tomorrow: https://github.com/flutter/flutter/issues/41881

**Pull request merged**